src/library/dao/trackdao.cpp

```diff
@@ -1545,143 +1548,162 @@ bool TrackDAO::detectMovedTracks(QSet<TrackId>* pTracksMovedSetOld,
+        case 1:
+            DEBUG_ASSERT(newTrackId.isValid());
+            DEBUG_ASSERT(newTrackLocationId.isValid());
+            kLogger.info()
+                    << "Found moved track location:"
+                    << oldTrackLocation
+                    << "->"
+                    << newTrackLocation;
+            break;
+        default:
+            kLogger.warning()
+                    << "Found too many ("
+                    << newTrackLocationCount
+                    << ") potential substitutes for missing track location"
+                    << oldTrackLocation;
+            continue;
         }
 
-        //If we found a moved track...
-        if (newTrackLocationId.isValid()) {
-            qDebug() << "Found moved track!" << filename;
-
-            // Remove old row from track_locations table
-            query.prepare("DELETE FROM track_locations WHERE id=:id");
-            query.bindValue(":id", oldTrackLocationId.toVariant());
-            if (!query.exec()) {
-                // Should not happen!
-                LOG_FAILED_QUERY(query);
-            }
+        TrackId oldTrackId(oldTrackQuery.value(oldTrackIdColumn));
+        DEBUG_ASSERT(oldTrackId.isValid());
+        DbId oldTrackLocationId(oldTrackQuery.value(oldLocationIdColumn));
+        DEBUG_ASSERT(oldTrackLocationId.isValid());
 
-            // The library scanner will have added a new row to the Library
-            // table which corresponds to the track in the new location. We need
-            // to remove that so we don't end up with two rows in the library
-            // table for the same track.
-            query.prepare("SELECT id FROM library WHERE location=:location");
-            query.bindValue(":location", newTrackLocationId.toVariant());
-            if (!query.exec()) {
-                // Should not happen!
-                LOG_FAILED_QUERY(query);
-            }
+        // The queries ensure that this assertion is always true (fs_deleted=0/1)!
+        DEBUG_ASSERT(oldTrackId != newTrackId);
 
-            const int query3idColumn = query.record().indexOf("id");
-            if (query.next()) {
-                TrackId newTrackId(query.value(query3idColumn));
-                query.prepare("DELETE FROM library WHERE id=:newid");
-                query.bindValue(":newid", newTrackLocationId.toVariant());
-                if (!query.exec()) {
-                    // Should not happen!
-                    LOG_FAILED_QUERY(query);
-                }
-                // We collect all the new tracks the where added to BaseTrackCache as well
-                pTracksMovedSetNew->insert(newTrackId);
-            }
-            // Delete the track
+        // The library scanner will have added a new row to the Library
+        // table which corresponds to the track in the new location. We need
+        // to remove that so we don't end up with two rows in the library
+        // table for the same track.
+        {
```

Comment:
What is if we found a duplicate and the track was already part of playlists?

Reply:
The newly added track can safely be deleted, because it is not referenced in any playlist. The track that was marked as missing will become visible again after referencing the new, existing location.

Reply:
But if we delete it here and something goes wrong below, the track is lost until the directory hash changes.